src/emulator/functionsEmulator.ts

```diff
@@ -818,7 +832,127 @@ export class FunctionsEmulator implements EmulatorInstance {
     return process.execPath;
   }
 
-  invokeRuntime(frb: FunctionsRuntimeBundle, opts: InvokeRuntimeOpts): RuntimeWorker {
+  getUserEnvs(): Record<string, string> {
+    const projectInfo = {
+      functionsSource: this.args.functionsDir,
+      projectId: "local", // "local" is reserved for Emulator use in functions env var.
+    };
+
+    if (functionsEnv.hasUserEnvs(projectInfo)) {
+      try {
+        return functionsEnv.loadUserEnvs(projectInfo);
+      } catch (e) {
+        // Ignore - user envs are optional.
+        logger.debug("Failed to load local environment variables", e);
+      }
+    }
+    return {};
+  }
+
+  getSystemEnvs(triggerDef?: {
+    targetName: string;
+    triggerType: EmulatedTriggerType;
+  }): Record<string, string> {
+    const envs: Record<string, string> = {};
+
+    // Env vars guaranteed by GCF platform.
+    //   https://cloud.google.com/functions/docs/env-var
+    envs.GCLOUD_PROJECT = this.args.projectId;
+    envs.K_REVISION = "1";
+    envs.PORT = "80";
+
+    if (triggerDef) {
+      const service = triggerDef.targetName;
+      const target = service.replace(/-/g, ".");
+      const mode = triggerDef.triggerType === EmulatedTriggerType.BACKGROUND ? "event" : "http";
+      envs.FUNCTION_TARGET = target;
+      envs.FUNCTION_SIGNATURE_TYPE = mode;
+      envs.K_SERVICE = service;
+    }
+    return envs;
+  }
+
+  getEmulatorEnvs(): Record<string, string> {
+    const envs: Record<string, string> = {};
+
+    envs.FUNCTIONS_EMULATOR = "true";
+    envs.TZ = "UTC"; // Fixes https://github.com/firebase/firebase-tools/issues/2253
+
+    // Make firebase-admin point at the Firestore emulator
+    const firestoreEmulator = this.getEmulatorInfo(Emulators.FIRESTORE);
+    if (firestoreEmulator != null) {
+      envs[Constants.FIRESTORE_EMULATOR_HOST] = formatHost(firestoreEmulator);
+    }
+
+    // Make firebase-admin point at the Database emulator
+    const databaseEmulator = this.getEmulatorInfo(Emulators.DATABASE);
+    if (databaseEmulator) {
+      envs[Constants.FIREBASE_DATABASE_EMULATOR_HOST] = formatHost(databaseEmulator);
+    }
+
+    // Make firebase-admin point at the Auth emulator
+    const authEmulator = this.getEmulatorInfo(Emulators.AUTH);
+    if (authEmulator) {
+      envs[Constants.FIREBASE_AUTH_EMULATOR_HOST] = formatHost(authEmulator);
+    }
+
+    // Make firebase-admin point at the Storage emulator
+    const storageEmulator = this.getEmulatorInfo(Emulators.STORAGE);
+    if (storageEmulator) {
+      envs[Constants.FIREBASE_STORAGE_EMULATOR_HOST] = formatHost(storageEmulator);
+      envs[Constants.CLOUD_STORAGE_EMULATOR_HOST] = `http://${formatHost(storageEmulator)}`;
```

Comment:
We should not need both of these, they were put in for backwards compat with earlier @google-cloud/storage SDKs, we can probably just do the Firebase one as long as we require a new-ish admin SDK

Reply:
That's good to know. Since this is already a "risky" refactor, I'll lean on making cleanup changes like that in a separate PR.